lib/Sema/CSApply.cpp

```diff
@@ -2514,6 +2514,65 @@ namespace {
             /*implicit=*/expr->isImplicit(), Type(), getType);
         result = finishApply(apply, Type(), cs.getConstraintLocator(expr));
       }
+			
+      // If the base type is optional and we're referring to an nominal
+      // type member via the dot syntax and the member name matches
+      // Optional<T>.none then emit a diagnostic
+      if (auto baseUnwrappedType = baseTy->lookThroughAllOptionalTypes()) {
+        if (auto DSCE = dyn_cast<DotSyntaxCallExpr>(result)) {
+          
+          Decl *structOrEnumMember = nullptr;
+          StringRef memberName;
+          auto calledValue = DSCE->getCalledValue();
+          bool isOptional = false;
+          
+          if (auto EED = dyn_cast<EnumElementDecl>(calledValue)) {
```

Comment:
I'm wondering if you actually need this whole if/else block at all. `ValueDecl` has `isInstanceMember()`, `getFullName()` (which would be a compound name if the member took parameters), and `getInterfaceType()`. Those *might* be enough to handle all of these checks without worrying about whether you're looking at an enum case or a constant.

Reply:
I tried using `ValueDecl` but it's tricky to check if the RHS is an optional using `getInterfaceType()` as it returns a `GenericFunctionType`. I couldn't really find a way to dig into it to check if the type is optional or not as `getOptionalObjectType()` doesn't work.

Anyway, I realised I don't actually have to check for a `VarDecl` here, just an `EnumElementDecl` as the compiler always favours `Optional<Foo>.caseName` over `Foo.caseName`, so we just have to check if `.caseName` is an `EnumElementDecl` and if it's `Optional`!

We can make the the check for `isInstanceMember()` once we get the result from lookup.